homeassistant/components/matrix/__init__.py

```diff
@@ -115,284 +139,274 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-        # join it twice by accident. However, fetching aliases is costly,
-        # so we only do it once per room.
-        self._aliases_fetched_for = set()
-
-        # Word commands are stored dict-of-dict: First dict indexes by room ID
-        #  / alias, second dict indexes by the word
-        self._word_commands = {}
+        self._client = AsyncClient(
+            homeserver=self._homeserver, user=self._mx_id, ssl=self._verify_tls
+        )
 
-        # Regular expression commands are stored as a list of commands per
-        # room, i.e., a dict-of-list
-        self._expression_commands = {}
+        self._listening_rooms = listening_rooms
 
-        for command in commands:
-            if not command.get(CONF_ROOMS):
-                command[CONF_ROOMS] = listening_rooms
-
-            if command.get(CONF_WORD):
-                for room_id in command[CONF_ROOMS]:
-                    if room_id not in self._word_commands:
-                        self._word_commands[room_id] = {}
-                    self._word_commands[room_id][command[CONF_WORD]] = command
-            else:
-                for room_id in command[CONF_ROOMS]:
-                    if room_id not in self._expression_commands:
-                        self._expression_commands[room_id] = []
-                    self._expression_commands[room_id].append(command)
+        self._word_commands: dict[RoomID, dict[WordCommand, ConfigCommand]] = {}
+        self._expression_commands: dict[RoomID, list[ConfigCommand]] = {}
+        self._load_commands(commands)
 
-        # Log in. This raises a MatrixRequestError if login is unsuccessful
-        self._client = self._login()
+        async def stop_client(event: HassEvent) -> None:
+            """Run once when Home Assistant stops."""
+            if self._client is not None:
+                await self._client.close()
 
-        def handle_matrix_exception(exception):
-            """Handle exceptions raised inside the Matrix SDK."""
-            _LOGGER.error("Matrix exception:\n %s", str(exception))
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_client)
 
-        self._client.start_listener_thread(exception_handler=handle_matrix_exception)
+        async def handle_startup(event: HassEvent) -> None:
+            """Run once when Home Assistant finished startup."""
+            self._auth_tokens = await self._get_auth_tokens()
+            await self._login()
+            await self._join_rooms()
+            # Sync once so that we don't respond to past events.
+            await self._client.sync(timeout=30_000)
 
-        def stop_client(_):
-            """Run once when Home Assistant stops."""
-            self._client.stop_listener_thread()
+            self._client.add_event_callback(self._handle_room_message, RoomMessageText)
 
-        self.hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_client)
+            await self._client.sync_forever(timeout=30_000)  # milliseconds.
```

Comment:
While testing this (works great!) I found this integration spamming my homeserver with sync requests as fast as it could. Would be great if this could be avoided, for example using `loop_sleep_time` argument for `sync_forever`.

Thanks for making this uplift, I could not get the previous implantation to work.

Reply:
Interesting, will check that out! Definitely may also be reasonable to allow `loop_sleep_time` to be configurable.